components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaClient.kt

```diff
@@ -0,0 +1,423 @@
+     *     Disconnect being executed at the same time from different threads.
+     *   - Application events are also sent serially. If one action causes an [AUTH_CHECK_STARTED]
+     *     state transition change, then the next causes [FxaAuthState.Disconnected], the
+     *     [FxaEventHandler.onStateChange] callback for the second change won't be called until
+     *     after the callback for the first change returns. This allows applications to ensure that
+     *     their UI reflects the current state.
+     *   - Actions are retried in the face of network errors and checkAuthorizationStatus is called on
+     *     authorization errors.  This allows the Rust client to recover when possible, for example
+     *     from expired access tokens when it holds a valid refresh token.
+     */
+    fun queueAction(action: FxaAction) {
+        actionProcessor.queue(action)
+    }
+
+    // This handles queueAction for us
+    private val actionProcessor = FxaActionProcessor(inner, eventHandler, { tryPersistState() })
+    init {
+        runActionProcessorManager(actionProcessor, coroutineContext)
+    }
+
+    /**
+     * Get the current authentication state
+     *
+     * FIXME: https://github.com/mozilla/application-services/issues/5819
+     */
+    fun getAuthState(): FxaAuthState = FxaAuthState.fromRust(inner.getAuthState())
+
+    // Wraps method calls that don't change the state, like [getAccessToken]
+    private suspend fun<T> wrapMethodCall(name: String, methodCall: () -> T): T {
+        Log.d(LOG_TAG, "Running: $name")
+        return withContext(coroutineContext) {
+            try {
+                methodCall()
+            } catch (e: FxaException.Authentication) {
+                queueAction(FxaAction.CheckAuthorization)
+                throw e
+            }
+        }
+    }
+
+    // Does what wrapMethodCall does and also ensures that the state is perisisted at the end of the
+    // operation.
+    private suspend fun<T> wrapMethodCallAndPersist(name: String, methodCall: () -> T): T {
+        return try {
+            wrapMethodCall(name, methodCall)
+        } finally {
+            // Use a finally block since we want to perisist the state regardless of if the method
+            // succeeded or not.
+            tryPersistState()
+        }
+    }
+
+    /**
+     * Fetches the profile object for the current client either from the existing cached account,
+     * or from the server (requires the client to have access to the profile scope).
+     *
+     * @param ignoreCache Fetch the profile information directly from the server
+     * @return [Profile] representing the user's basic profile info
+     * @throws FxaException.Unauthorized We couldn't find any suitable access token to make that call.
+     * The caller should then start the OAuth Flow again with the "profile" scope.
+     */
+    suspend fun getProfile(ignoreCache: Boolean): Profile = wrapMethodCallAndPersist("getProfile") {
```

Comment:
If `ignoreCache` is false and we end up returning a cached value, we probably don't need to persist anything?